activate

```diff
@@ -1,95 +1,218 @@
-json="$(python -c 'import json, os, sys; print json.dumps(dict((v,os.environ[v]) for v in sys.argv[1:]))' ${vars})"
-
-# report in
-if ! out=$(wget -O - -q --method=PUT --body-data="${json}" "${SCRIPT}") ; then
-    echo "failed PUT to $SCRIPT with data: ${json}"
-    exit 1
-fi
-if ! ( echo "${out}" | python -c "import json, sys; sys.exit(0 if json.load(sys.stdin)['success'] == 'true' else 1)" 2>/dev/null ); then
-    echo "didn't get success from PUT to ${SCRIPT} with data: ${json}"
-    echo "got response: ${out}"
-    exit 1
-fi
-
-# touch file to record activation
-echo > "${ACTIVATEDFILE}"
+#!/usr/bin/env python3
 
+import json
+import os
+import re
+import subprocess
+import sys
+import urllib.request
+
+class Activation(object):
+    ACTIVATION_ENDPOINT = os.getenv('EOS_PHONE_HOME_ACTIVATE_URL', "https://home.endlessm.com/v1/activate")
+    STATE_DIRECTORY = "/var/lib/eos-phone-home"
+
+    ACTIVATION_VARIABLES = { 'live': 'is_live',
+                             'image': 'image',
+                             'release': 'release',
+                             'vendor': 'vendor',
+                             'product': 'product',
+                             'serial': 'serial' }
+
+    def _get_vendor(self):
+        vendor = None
+
+        try:
+            if os.path.isdir('/sys/class/dmi/id'):
+                with open('/sys/class/dmi/id/sys_vendor') as vendor_file:
+                    vendor = vendor_file.read().strip()
+            else:
+                dt_data = None
+                with open('/proc/device-tree/compatible') as dt_file:
+                    dt_data = dt_file.read().strip()
+
+                dt_fields = dt_data.split(',')
+                vendor = dt_fields[0]
+        except Exception as e:
+            print("Unable to get vendor name!", e, file=sys.stderr)
+
+        print(" - Found vendor:", vendor)
+
+        return vendor.strip()
+
+    def _get_product(self):
+        product = None
+
+        try:
+            if os.path.isdir('/sys/class/dmi/id'):
+                with open('/sys/class/dmi/id/product_name') as product_file:
+                    product = product_file.read().strip()
+            else:
+                dt_data = None
+                with open('/proc/device-tree/compatible') as dt_file:
+                    dt_data = dt_file.read().strip()
+
+                dt_fields = dt_data.split(',')
+                product = dt_fields[1]
+        except Exception as e:
+            print("Unable to get product name!", e, file=sys.stderr)
+
+        print(" - Found product:", product)
+
+        return product.strip()
+
+    def _get_image(self):
+        image = None
+
+        try:
+            image = subprocess.check_output(['attr', '-q', '-g', 'eos-image-version', '/sysroot']).strip()
+        except Exception as e:
+            print("Unable to get /sysroot image name!", e, file=sys.stderr)
+            return None
+
+        print(" - Image:", image)
+
+        if image != "":
```

Comment:
The logic here confused me on first read, and I'm not sure it's right - if the attr on /sysroot is not set (ie returns "") we will return None instead of "" and won't check on /. Considering the likely case is a non-converted system, where the attribute is on /sysroot, I'd go for an early return as more efficient/readable, only check / in the fallback case:

image=""

try:
  get attr from /sysroot

if image != "":
  return image

try:
  get attr from /

return image

(or if you're very keen, return image == "" ? None : image)